grpc/creds/creds.go

```diff
@@ -51,9 +63,73 @@ func (tc *transportCredentials) ClientHandshake(ctx context.Context, addr string
 	}
 }
 
+func (tc *transportCredentials) peerIsWhitelisted(peerState tls.ConnectionState) error {
+	// If there's no whitelist, all clients are OK
+	if tc.whitelist == nil {
+		return nil
+	}
+
+	// Otherwise its time to start inspecting the peer's `VerifiedChains`
+	chains := peerState.VerifiedChains
+	if len(chains) < 1 {
```

Comment:
We never get here if no trusted path can be built. I think you might be right that the chains will all start with the same leaf certificate in retrospect.

I'll double check the upstream `tls` source and revisit.